### PR TITLE
Add link color support to Nav block to allow for override of Theme JSON styles

### LINKED ISSUE #41146: Navigation block: text color not appearing in the editor
### Description

When assigning a text color to the main navigation block using TT2, the color isn't shown properly in the editor but does appear correctly when viewed on the front end of the site. This occurs with 6.0 RC3. cc @ndiego as a heads up, @getdave to perhaps help address, and @scruffian for possibility of this being a TT2 issue. 

### Step-by-step reproduction instructions

1. Add a navigation block.
2. Select a text color.
3. Notice no changes in editor.
4. Publish changes.
5. View frontend of the site and see changes there.

### Screenshots, screen recording, code snippet

https://user-images.githubusercontent.com/26996883/169198201-0897ec69-5b83-4584-92cf-dddb746a9dff.mov

### Environment info

- WordPress 6.0 RC 3
- TT2
- No GB installed

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by getdave**
## What?
Fixes https://github.com/WordPress/gutenberg/issues/41146 by adding `linkColor` support to the Nav block thereby ensuring that Nav block specific link color styles can take precedence over inherited Theme JSON styles. 

## Why?

In https://github.com/WordPress/gutenberg/issues/41146#issuecomment-1132775573 we learnt that the Theme JSON styles coming from _link_ color styles on the enclosing Group block (the one wrapping the Nav block) were taking precedence over the user selected text color styles on the Nav block.

This is correct (if unintuative) behaviour as CSS cascade means the link selector from the Global Styles will target the `<a>` of the Nav block. That overrides the text color set on the Nav block.

This is becayse textColor !== linkColor.

We need to provide a way for the Nav block to overide linkColor styles cascading from Theme JSON.

## How?

This PR:

- manually (_not_ via block supports) adds `linkColor` and `customLinkColor` support to the Navigation block.
- ensures the values of the above are passed via context and consumed by 
    - Navigation Submenu block
    - Navigation Link block
- adds appropriate classes to the aforesaid blocks
- ensures that CSS rules and specificity is adjust to ensure that the link color is `inherit`ed from the appropriate parent on to the appropriate element.

The above requires jumping through many hoops and thus why there are so many file changes.

    

## Alternatives considered

Initially this PR targetted the existing `textColor` on the Nav block and ensured that took precedence on any styles cascading from Global Styles. However as @scruffian [pointed out that is counterintuitive](https://github.com/WordPress/gutenberg/pull/41199#issuecomment-1132841240) as link and text color are not the same thing.

One concern here is that for a while Nav block consumers have probably seen the `Text` color control as equivalent to "the color of the links in the Nav". However in the future we may have text items in a Nav which are not links and so we need to start to differentiate them.

## Testing Instructions

- Add a Nav block and some items.
- Wrap the Nav in a Group and set a link color on that group.
- See that by default the Nav inherits the link color set on the Group
- Save and check the front end also respects the above.
- Now set a specific Link Color on the Nav block which is different to that on the Nav block.
- See that this color is respected.
- Save and check the front end also respects the above.
- Now set an `Overlay and submenu link` color on the Nav block. Make sure this distinct from the previous color and the one on the Group block. 
- Check this renders correctly in Editor and front end.
- Remove the link color from the Nav block, whilst retaining the submenu link color. Again check things inherit correctly.
- Try other variations of the above to ensure CSS cascades correctly.
- Check the overlay also obeys these rules.

🤔 We should also test that adding styles from Theme JSON that _specifically_ target the Nav block are still respected.

## Screenshots or screencast 

https://user-images.githubusercontent.com/444434/169521486-8d1a8202-52fc-4198-83de-a14c1604c8e4.mp4

### CODE REVIEW

**Comment by scruffian** [comment]
I'm not sure about this approach. It seems to me that the behaviour in the editor is the correct behaviour. In CSS if I set a link color in an element then to override it I need to set a different _link_ color in the child - setting a text color in the child won't be enough. 

I think what we need to do is to make it possible to set a link color in the nav block, rather than a text color. This is a bit confusing though as I'm not sure its possible to have text in a navigation block which isn't a link, although one day it probably will be.

**Comment by getdave** [comment]
> I'm not sure about this approach. It seems to me that the behaviour in the editor is the correct behaviour. In CSS if I set a link color in an element then to override it I need to set a different _link_ color in the child - setting a text color in the child won't be enough.
> 
> I think what we need to do is to make it possible to set a link color in the nav block, rather than a text color. This is a bit confusing though as I'm not sure its possible to have text in a navigation block which isn't a link, although one day it probably will be.

Yeh you're right. It's the link color not the text color that's overwriting the block specific styles. Ok let's add linkcolor support to the Nav block.

**Comment by getdave** [comment]
@scruffian I added `linkColor` support to the Nav block and adjusted the CSS selector to `has-link-color`.

**Comment by getdave** [comment]
@scruffian I've noticed the CSS specificity is different in the editor on the Global Styles selector due to it being prepended with `.editor-styles-wrapper`. That's causing the inherit behaviour to be off on the front end.

File: packages/block-library/src/navigation/style.scss [hunk 1]
```diff
@@ -41,8 +41,6 @@ $navigation-icon-size: 24px;
 	}
 
 	// Menu item link.
-	// By adding low specificity, we enable compatibility with link colors set in theme.json,
-	// but still allow them to be overridden by user-set colors.
 	.wp-block-navigation-item__content {
 		color: inherit;
```
Comment (scruffian) [inline packages/block-library/src/navigation/style.scss:45]: I think we should remove this line. `.wp-block-navigation-item__content` is a class that is used on a link. This line forces links to inherit their color from their parent which is against the pattern of CSS.

**Comment by scruffian** [comment]
Sorry, I'm not sure the approach here is the right one. The elements API expects the link color to be stored in a very specific attribute:

https://github.com/WordPress/gutenberg/blob/d36ee9642b5d64e08cf088839dd3179d8d310429/lib/block-supports/elements.php#L98

If we do this then we won't need to write any custom code to apply the colors, as the library should take care of that for us.

**Comment by getdave** [comment]
> If we do this then we won't need to write any custom code to apply the colors, as the library should take care of that for us.

I'm pretty sure the Nav block doesn't follow "the rules" here. All the stuff for text and background color is custom coded. I think this is because the colours aren't intended for the Nav block itself but rather to be inherited by the child blocks.

I can try uses the standard approach you suggest but I deliberately avoided this and followed the patterns that have already been established in the block (e.g. `textColor`).

**Comment by getdave** [comment]
Some more research reveals that the need for custom colors for the `Submenu and Overlay` is probably the reason why we have a custom `colors` implementation here.

If I add support for `linkColor` using the "standards" way via `block.json` 

```json
"supports": {
    "color": {
        "link": true
    }
}
```

...then I get x2 color controls:

Screen Shot 2022-05-23 at 09 18 24

In the longer term we may need a wider refactor of how Submenu and Overlay colors are handled. For now though in order to fix this bug I think building on the existing custom implementation would be fine.

**Comment by getdave** [comment]
Other options:

## Separate Submenu & Overlay colors panel

We can use standard block supports for the text, link and background colors.

Then we can have a dedicated panel for the "custom" color controls for Submenu and Overlay.

Screen Shot 2022-05-23 at 09 31 45

**Comment by scruffian** [comment]
Ok, this one is going to get complicated. My suggestion is:
- Ship this (or something like it) to fix the bug
- Look into what an ideal solution might look like...

File: packages/block-library/src/navigation-link/block.json [hunk 2]
```diff
@@ -43,10 +43,14 @@
 	"usesContext": [
 		"textColor",
 		"customTextColor",
+		"linkColor",
+		"customLinlColor",
```
Comment (draganescu) [inline packages/block-library/src/navigation-link/block.json:47]:
```suggestion
		"customLinkColor",
```

**Review by draganescu — CHANGES_REQUESTED** [review]
I believe this is a good opportunity to not patch to fix. If I am not mistaken we'd be introducing things that we'll need to remove later.

We already have a system in place and we should try to improve that one for this specific use case:

- properties set on parents should be inherited by children, unless children override them

As far as I understand the global styles system supports this just fine, but the navigation block has a custom implementation. Maybe it's time to remove that and make sure we fit into what the platform currently offers? Probably at the time the global styles system did not offer us a way to implement what we wanted, or whatever was the reason.

I understand that refactoring is a big undertaking and the bug is now, but fixing will require us to support temporary solutions and remove them later, they're not even experimental. Also, I believe the fix now refactor later should only apply to core on critical bugs. Is this color inheritance on one block a critical bug?

**Comment by getdave** [comment]
Thanks for your review @draganescu . It's much appreciated.

> I believe this is a good opportunity to not patch to fix. If I am not mistaken we'd be introducing things that we'll need to remove later.

You are correct. The question is, how much are we really compounding the problem vs being able to move quickly to patch a fix. 

> As far as I understand the global styles system supports this just fine, but the navigation block has a custom implementation.
> Probably at the time the global styles system did not offer us a way to implement what we wanted, or whatever was the reason.

Global Styles has block supports for `color` which supports text, link and background. However there is no way to add custom colors AFAIK.

The Nav block has a custom implementation for these things because we need to support custom color setitngs for overlay and submenu as per https://github.com/WordPress/gutenberg/pull/41199#issuecomment-1134351817.

> I understand that refactoring is a big undertaking and the bug is now, but fixing will require us to support temporary solutions and remove them later, they're not even experimental. Also, I believe the fix now refactor later should only apply to core on critical bugs. Is this color inheritance on one block a critical bug?

I think we owe it to the original reporters and our Themes community to check on the impact of this bug. If it's something that can be "managed" we can go straight to the refactor. If it's critical then we'll need to patch and then refactor.

I would also prefer _**not**_ to compound the existing tech debt 👍

**Comment by getdave** [comment]
Closing this one as I have a new plan. I will raise a PR shortly.

**Pull request closed without merging**